core/src/typecheck/error.rs

```diff
@@ -112,8 +112,9 @@ impl UnifError {
     ///
     /// Wrapper that calls [`Self::into_typecheck_err_`] with an empty [name
     /// registry][reporting::NameReg].
-    pub fn into_typecheck_err(self, state: &State, pos_opt: TermPos) -> TypecheckError {
-        self.into_typecheck_err_(state, &mut reporting::NameReg::new(), pos_opt)
+    pub fn into_typecheck_err(self, state: &mut State, pos_opt: TermPos) -> TypecheckError {
+        let mut names = reporting::NameReg::new(std::mem::take(state.names));
+        self.into_typecheck_err_(state, &mut names, pos_opt)
```

Comment:
This function now replaces the `NameReg` in `state` by an empty one and calls `into_typecheck_err_` with the old, filled one. This might end up being very surprising. I wouldn't expect that `into_typecheck_err` removes name allocations from the state. Maybe it would make more sense to take ownership of `state`?

In any case, the doc-comment needs to be updated.

Reply:
Good point, but doing that at the caller site might be very annoying, because `State` is taken by reference everywhere, so we would end up with a lot of `std::mem::take`.

In practice, when you call to `into_typecheck_err`, you go the error path, so you don't need the state anymore. But you're right that the behavior isn't necessarily natural. Maybe cloning is the path of least resistance. Another solution is to separate the name tables between the one coming from state, that is would be immutably borrowed, and the one corresponding to generated names, that is owned, but that also sounds like more boilerplate (lifetimes, wrapper around `get`, etc.) just to avoid one clone.

Reply:
I'll just clone, it's the simplest.